versioned_hdf5/tests/test_slicetools.py

```diff
@@ -36,3 +44,400 @@ def test_spaceid_to_slice(h5file):
+def test_read_many_slices_noncontiguous_idx(h5file, use_h5, contiguous_cols):
+    """one or more coordinate array is not C-contiguous along the rows or both
+    columns and rows.
+
+    Non-contiguous rows are supported thanks to Cython strided views.
+    Non-contiguous columns are not supported in hdf5 fast mode and are transparently
+    deep-copied before use.
+    """
+    src = np.arange(1, 51).reshape(10, 5)
+
+    src_start = np.array([(0, 1, 2), (9, 9, 9), (2, 3, 4)])
+    if contiguous_cols:
+        src_start = src_start[::2, :2]
+        expect = [[src[0, 1]], [src[2, 3]]]
+    else:
+        src_start = src_start[::2, ::2]
+        expect = [[src[0, 2]], [src[2, 4]]]
+
+    assert not src_start.flags.c_contiguous
+
+    if use_h5:
+        src = h5file.create_dataset("a", data=src)
+
+    dst = np.zeros((2, 1), dtype=src.dtype)
+    read_many_slices(src, dst, src_start, [(0, 0), (1, 0)], (1, 1))
+    np.testing.assert_equal(dst, expect)
+
+
+@pytest.mark.parametrize("use_h5", [True, False])
+def test_read_many_slices_src_size_zero(h5file, use_h5):
+    """src array has size 0"""
+    src = np.empty((3, 0))
+    dst = np.zeros((3, 3), dtype=src.dtype)
+    if use_h5:
+        src = h5file.create_dataset("a", data=src)
+    read_many_slices(src, dst, [(0, 0)], [(0, 0)], [(3, 3)])
+    assert (dst == 0).all()
+
+
+@pytest.mark.parametrize("use_h5", [True, False])
+def test_read_many_slices_dst_size_zero(h5file, use_h5):
+    """dst array has size 0"""
+    src = np.arange(1, 10).reshape(3, 3)
+    dst = np.empty((0, 3), dtype=src.dtype)
+    if use_h5:
+        src = h5file.create_dataset("a", data=src)
+    read_many_slices(src, dst, [(0, 0)], [(0, 0)], [(3, 3)])
+
+
+def test_read_many_slices_noncontiguous_dst(h5file):
+    """dst is a numpy array that's not C-contiguous.
+    It is supported, but only with fast=False.
+    """
+    src = np.arange(1, 17).reshape(4, 4)
+    expect = np.zeros((8, 12), dtype=src.dtype)
+    expect[::2, ::3] = src[()]
+
+    dst = np.zeros_like(expect)
+    read_many_slices(src, dst[::2, ::3], [(0, 0), (2, 0)], [(0, 0), (2, 0)], (2, 4))
+    np.testing.assert_equal(dst, expect)
+
+    src = h5file.create_dataset("a", data=src)
+    dst = np.zeros_like(expect)
+    read_many_slices(
+        src, dst[::2, ::3], [(0, 0), (2, 0)], [(0, 0), (2, 0)], (2, 4), fast=False
+    )
+    np.testing.assert_equal(dst, expect)
+
+    # h5py src only works in slow mode
+    with pytest.raises(NotImplementedError, match="contiguous"):
+        read_many_slices(
+            src, dst[::2, ::3], [(0, 0), (2, 0)], [(0, 0), (2, 0)], (2, 4), fast=True
+        )
+
+    # For the sake of awareness, user must explicitly pass fast=False
+    with pytest.raises(NotImplementedError, match="contiguous"):
+        read_many_slices(src, dst[::2, ::3], [(0, 0), (2, 0)], [(0, 0), (2, 0)], (2, 4))
+
+
+def test_read_many_slices_not_fast_read_ok(h5file):
+    """src is a h5py dataset that doesn't support fast read"""
```

Comment:
It might be good to outline somewhere in the docs when a dataset supports fast reading, since there is a significant difference in performance.